### PR TITLE
Use `max_radius` in TRON

### PR DESCRIPTION
**Pull request by tmigot**
#181

### CODE REVIEW

**Comment by tmigot** [comment]
@abelsiqueira Making a release after this one, should solve the use of `max_radius` in Percival.

**Comment by abelsiqueira** [comment]
I think this should be tested with the new changes to SolverTools

File: src/tron.jl [hunk 1]
```diff
@@ -109,31 +113,30 @@ function TronSolver(nlp::AbstractNLPModel{T, V};) where {T, V <: AbstractVector{
...
   H = hess_op!(nlp, x, Hs)
   Op = typeof(H)
-  tr = TrustRegion(gt, one(T))
+  tr = TRONTrustRegion(gt, min(one(T), max_radius - eps(T)); max_radius = max_radius, kwargs...)
   return TronSolver{T, V, Op}(x, xc, temp, gx, gt, gn, gpx, Hs, H, tr)
 end
 
 function SolverCore.reset!(solver::TronSolver)
   solver.tr.good_grad = false
-  solver.tr.radius = solver.tr.initial_radius
```
Comment (abelsiqueira) [inline src/tron.jl:118]: Why remove this?

File: src/tron.jl [hunk 2]
```diff
@@ -94,10 +94,14 @@ mutable struct TronSolver{T, V <: AbstractVector{T}, Op <: AbstractLinearOperato
...
   Hs::V
   H::Op
-  tr::TrustRegion{T, V}
+  tr::TRONTrustRegion{T, V}
 end
 
-function TronSolver(nlp::AbstractNLPModel{T, V};) where {T, V <: AbstractVector{T}}
+function TronSolver(
+  nlp::AbstractNLPModel{T, V};
+  max_radius::T = min(one(T) / sqrt(2 * eps(T)), T(100)),
```
Comment (abelsiqueira) [inline src/tron.jl:102]: We are disassociating the `max_radius` from the `initial_radius`, but maybe it is best if we accept an additional argument, for `initial_radius`?
Or maybe we should accept `trust_region_kwargs`, like in Percival? We can add `TR <: AbstractTrustRegion{T, V}`.

File: test/runtests.jl [hunk 3]
```diff
@@ -17,3 +17,16 @@ if VERSION ≥ v"1.6"
...
+  max_radius = 0.00314
+  function cb(nlp, solver, stats)
+    @test solver.tr.radius ≤ max_radius
+  end
+
+  nlp = ADNLPModel(x -> 100 * (x[2] - x[1]^2)^2 + (x[1] - 1)^2, [-1.2; 1.0])
+  stats = tron(nlp, max_radius = max_radius, callback = cb)
+
+  nls = ADNLSModel(x -> [100 * (x[2] - x[1]^2); x[1] - 1], [-1.2; 1.0], 2)
+  stats = tron(nls, max_radius = max_radius, callback = cb)
```
Comment (abelsiqueira) [inline test/runtests.jl:31]: Nice test

**Review by abelsiqueira — CHANGES_REQUESTED** [review]
Overall nice. I think that maybe we need to allow changing the initial radius as well. It is possible that this "feature" os `max_radius` setting the initial radius was used at some point.
Even more generic would be using `AbstractTrustRegion` instead and passing kwargs to it, but that can be a new issue.

File: src/tron.jl (hunk 1 above)
Comment (tmigot) [reply src/tron.jl:118]: Because, it is done in the `solve!` function now

File: src/tron.jl (hunk 2 above)
Comment (tmigot) [reply src/tron.jl:103]: I added a mechanism so that all the kwargs taken by `TRONTrustRegion` are valid now.

**Comment by tmigot** [comment]
@abelsiqueira I also made a branch in Percival.jl anticipating these changes https://github.com/JuliaSmoothOptimizers/Percival.jl/compare/add-max-radius

File: src/tron.jl (hunk 2 above)
Comment (abelsiqueira) [reply src/tron.jl:103]: It is good until generalization

**Comment by abelsiqueira** [comment]
I didn't check if the tests used the latest SolverTools

**Comment by abelsiqueira** [comment]
They did, just checked.

**Pull request merged**